File: discussion/remove_if.py
def crit(x):
    return x > 10


def remove_if(arr, crit):
    '''Remove elements in place'''
    good = 0 # good arrays
    for i in range(len(arr)):
        if crit(arr[i]):
            arr[good] = arr[i]
            good += 1
    return arr[:good]


if __name__ == "__main__":
    arr = [i for i in range(20)]
    print("Original: ", arr)
    print("Remove_if: ", remove_if(arr, crit))

File: discussion/remove_if.cpp
#include <vector>
using namespace std;

bool criteria(int x) {
    return (x > 10);
}

void remove_if(std::vector<int> &arr) {
    // Keep track of good elements
    int good = 0;
    for (int i = 0; i < size(arr); i++) {
        if (criteria(arr[i])) {
            if (i != good) {
                arr[good] = arr[i];
            }
            good++;
        }
    }
    arr.resize(good);
}

